### PR TITLE
stac-10915 servicenow tags

### PR DESCRIPTION
**Pull request by vishalxebia**
### Step 3: Did you add / update tests for your changes in the right area?
- [ ] Unit/Component test
- [ ] Integration test	

### Step 4: I'm confident that everything is properly tested:
I got a PO / QA Approval by:
- [ ] Name

### Step 5: Can we ship this feature to production?
- [ ] Yes, I'm proud of my work. Ship it! :ship:

### CODE REVIEW

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 1]
```diff
@@ -57,6 +62,27 @@ class ChangeRequest(Model):
...
+    name = ModelType(WrapperStringType, required=True)
+    sys_class_name = ModelType(WrapperStringType, required=True)
+    sys_id = ModelType(WrapperStringType, required=True)
+    sys_tags = ModelType(WrapperStringType, default=WrapperStringType())
+    fqdn = ModelType(WrapperStringType, default=WrapperStringType())
+    host_name = ModelType(WrapperStringType, default=WrapperStringType())
+
+
+class CIRelation(Model):
+    sys_id = ModelType(WrapperStringType)
```
Comment (zandrewitte) [inline servicenow/stackstate_checks/servicenow/servicenow.py:75]: Is `sys_id` required?

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 2]
```diff
@@ -209,52 +244,37 @@ def _process_components(self, instance_info):
         collected_components = self._batch_collect(self._batch_collect_components, instance_info)
 
         for component in collected_components:
+            try:
+                config_item = ConfigurationItem(component, strict=False)
+                config_item.validate()
+            except DataError as e:
+                self.log.warning("Error while processing properties - {}".format(e))
```
Comment (zandrewitte) [inline servicenow/stackstate_checks/servicenow/servicenow.py:251]: Add the component `sys_id` and `name` to the log

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 3]
```diff
@@ -145,17 +173,24 @@ def _get_sys_class_relation_filter_query(self, sys_class_filter):
         return sysparm_query
 
     @staticmethod
-    def _filter_empty_metadata(data):
+    def _apply_field_metadata(data, display_value_list):
```
Comment (zandrewitte) [inline servicenow/stackstate_checks/servicenow/servicenow.py:176]: This actually only selects `display_value` or `value` depending on whether it's in `display_value_list`. Can we maybe rename the function to `select_metadata_field`... or something better :smile:

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 4]
```diff
@@ -272,16 +292,28 @@ def _process_relations(self, instance_info):
...
-            type_sys_id = relation['type']['value']
-
-            relation_type = relation_types[type_sys_id]
-            data = self._filter_empty_metadata(relation)
-            data.update({"tags": instance_info.instance_tags})
+            try:
+                ci_relation = CIRelation(relation, strict=False)
+                ci_relation.validate()
+            except DataError as e:
+                self.log.warning("Error while processing properties - {}".format(e))
```
Comment (zandrewitte) [inline servicenow/stackstate_checks/servicenow/servicenow.py:301]: Add the relation `sys_id` ro the log

File: servicenow/stackstate_checks/servicenow/servicenow.py [hunk 5]
```diff
@@ -272,16 +292,28 @@ def _process_relations(self, instance_info):
...
+                continue
+            data = {}
+            relation = self._apply_field_metadata(relation, instance_info.relation_display_value_list)
+            parent_sys_id = ci_relation.parent.value
+            child_sys_id = ci_relation.child.value
+            # first part after splitting with :: contains actual relation
+            relation_type = ci_relation.type.display_value.split("::")[0]
```
Comment (zandrewitte) [inline servicenow/stackstate_checks/servicenow/servicenow.py:308]: Let's maybe add a check here to make sure the display value is not empty and that it contains a `::`, so something like:
```
if ci_relation.type.display_value and "::" in ci_relation.type.display_value
```

**Review by zandrewitte — CHANGES_REQUESTED** [review]

File: servicenow/stackstate_checks/servicenow/servicenow.py (hunk 1 above)
Comment (vishalxebia) [reply servicenow/stackstate_checks/servicenow/servicenow.py:75]: nope it's optional. we just need `parent id` and `child id`. We don't have any concept for `external_id` right now for relation. so not required.

File: servicenow/stackstate_checks/servicenow/servicenow.py (hunk 3 above)
Comment (vishalxebia) [reply servicenow/stackstate_checks/servicenow/servicenow.py:176]: yeah, I couldn't think better.

File: servicenow/stackstate_checks/servicenow/servicenow.py (hunk 5 above)
Comment (vishalxebia) [reply servicenow/stackstate_checks/servicenow/servicenow.py:308]: it's actually a required field and we do that using schematics already. should we validate again?

File: servicenow/stackstate_checks/servicenow/servicenow.py (hunk 4 above)
Comment (vishalxebia) [reply servicenow/stackstate_checks/servicenow/servicenow.py:301]: added.

File: servicenow/stackstate_checks/servicenow/servicenow.py (hunk 2 above)
Comment (vishalxebia) [reply servicenow/stackstate_checks/servicenow/servicenow.py:251]: added.

**Review by zandrewitte — APPROVED** [review]

**Pull request merged**